backend/routers/users.js

```diff
@@ -0,0 +1,144 @@
+      throw createError(400, 'Username is required.');
+    }
+    if (!password) {
+      throw createError(400, 'Password is required.');
+    }
+    if (!email) {
+      throw createError(400, 'Email is required.');
+    }
+    if (!isValidEmail(email)) {
+      throw createError(400, 'Email is not valid.');
+    }
+    if (password.length < 8) {
+      throw createError(400, 'Password must be at least 8 characters.');
+    }
+    if (await UserModel.findOne({ username })) {
+      throw createError(400, 'Username is already taken.');
+    }
+    if (await UserModel.findOne({ email })) {
+      throw createError(400, 'Email is already taken.');
+    }
+    const passwordHash = await bcrypt.hash(password, 8);
+    const user = await UserModel.create({
+      email,
+      username,
+      password: passwordHash,
+      kingdomName: kingdomName || `${username}'s kingdom`,
+      kingdomList: [],
+    });
+    user.password = undefined;
+    const token = jwt.sign({ user }, process.env.APP_SECRET || 'testSecret');
+    res.status(201).send({ token });
+  } catch (error) {
+    next(error);
+  }
+}
+
+async function login(req, res, next) {
+  const { username, password } = req.body;
+  try {
+    if (!username && !password) {
+      throw createError(400, 'All fields are required.');
+    }
+    if (!username) {
+      throw createError(400, 'Username is required.');
+    }
+    if (!password) {
+      throw createError(400, 'Password is required.');
+    }
+    const user = await UserModel.findOne({ username });
+    if (!user || !await bcrypt.compare(password, user.password)) {
+      throw createError(400, 'Username or password is incorrect.');
+    }
+
+    user.password = undefined;
+    const token = jwt.sign({ user }, process.env.APP_SECRET || 'testSecret');
+    res.status(200).send({ token });
+  } catch (error) {
+    next(error);
+  }
+}
+
+async function updateUserInfo(req, res, next) {
+  const { username: oldUsername } = req.user;
+  const {
+    username: newUsername, email: newEmail, password: newPassword, kingdomName: newKingdomName,
+  } = req.body;
+  try {
+    const changedTarget = {};
+    if (newUsername) {
+      changedTarget.username = newUsername;
+    }
+    if (newEmail) {
+      changedTarget.email = newEmail;
+    }
+    if (newPassword) {
+      changedTarget.password = newPassword;
+    }
+    if (newKingdomName) {
+      changedTarget.kingdomName = newKingdomName;
+    }
+    const changedValue = Object.keys(changedTarget);
+    if (changedValue.length === 0) {
+      throw createError(400, 'Please fill at least one element');
+    }
+    const user = await UserModel.findOneAndUpdate(
+      { username: oldUsername },
+      { $set: changedTarget },
+      { new: true, fields: '-_id' },
+    );
+    if (!user) {
+      throw createError(400, 'Can\'t find this username in database');
+    }
+    const message = `${changedValue.join(', ')} are successfully changed!`;
+    res.status(202).send({ message });
```

Comment:
I think I'll send this message back to the front as a toast. So that users will know which part already been changed.

Reply:
just sayin, that the text that is rendered (in the toast, or in an alert, or in a component) should be written on the frontend side based on the status code (except error messages), so it's ok now if you want to send it back, but in the future it's not the best practice.
also if the user changes only one thing, there should be an `is` instead of the `are` in the message